xbmc/music/ContextMenus.cpp

```diff
@@ -70,7 +72,14 @@ bool CMusicPlay::IsVisible(const CFileItem& item) const
 
 bool CMusicPlay::Execute(const std::shared_ptr<CFileItem>& item) const
 {
-  MUSIC_UTILS::PlayItem(item);
+  // auto play next item only works from inside certain media windows
+  MUSIC_UTILS::PlayMode mode = MUSIC_UTILS::PlayMode::PLAY_ONLY_THIS;
+  const int activeWindow = CServiceBroker::GetGUI()->GetWindowManager().GetActiveWindow();
```

Comment:
is it not possible to pass a property when the item is actually clicked/played in musicnav and use it to deduce the mode instead of checking the active window here?

Reply:
Good idea. Let's see whether I find a way to implement it in a clean way (at a central place in the music windows code).

Reply:
Ah wait, I have an idea now...

Reply:
@enen92 would this be something more acceptable for you?

```diff
diff --git a/xbmc/music/ContextMenus.cpp b/xbmc/music/ContextMenus.cpp
index d919609a3c..4c26ad374a 100644
--- a/xbmc/music/ContextMenus.cpp
+++ b/xbmc/music/ContextMenus.cpp
@@ -72,14 +72,12 @@ bool CMusicPlay::IsVisible(const CFileItem& item) const
 
 bool CMusicPlay::Execute(const std::shared_ptr<CFileItem>& item) const
 {
-  // auto play next item only works from inside certain media windows
-  ContentUtils::PlayMode mode = ContentUtils::PlayMode::PLAY_ONLY_THIS;
-  const int activeWindow = CServiceBroker::GetGUI()->GetWindowManager().GetActiveWindow();
-  if (activeWindow == WINDOW_MUSIC_NAV)
-    mode = ContentUtils::PlayMode::CHECK_AUTO_PLAY_NEXT_ITEM;
+  const ContentUtils::PlayMode mode = item->GetProperty("CheckAutoPlayNextItem").asBoolean()
+                                          ? ContentUtils::PlayMode::CHECK_AUTO_PLAY_NEXT_ITEM
+                                          : ContentUtils::PlayMode::PLAY_ONLY_THIS;
+  MUSIC_UTILS::PlayItem(item, mode);
 
   item->SetProperty("playlist_type_hint", PLAYLIST::TYPE_MUSIC);
-  MUSIC_UTILS::PlayItem(item, mode);
   return true;
 }
 
diff --git a/xbmc/music/windows/GUIWindowMusicNav.cpp b/xbmc/music/windows/GUIWindowMusicNav.cpp
index f8ba5a4549..a0294ab7d6 100644
--- a/xbmc/music/windows/GUIWindowMusicNav.cpp
+++ b/xbmc/music/windows/GUIWindowMusicNav.cpp
@@ -672,6 +672,17 @@ void CGUIWindowMusicNav::GetContextButtons(int itemNumber, CContextButtons &butt
   CGUIWindowMusicBase::GetNonContextButtons(buttons);
 }
 
+bool CGUIWindowMusicNav::OnPopupMenu(int iItem)
+{
+  if (iItem >= 0 && iItem < m_vecItems->Size())
+  {
+    const auto item = m_vecItems->Get(iItem);
+    item->SetProperty("CheckAutoPlayNextItem", {true});
+  }
+
+  return CGUIWindowMusicBase::OnPopupMenu(iItem);
+}
+
 bool CGUIWindowMusicNav::OnContextButton(int itemNumber, CONTEXT_BUTTON button)
 {
   CFileItemPtr item;
diff --git a/xbmc/music/windows/GUIWindowMusicNav.h b/xbmc/music/windows/GUIWindowMusicNav.h
index 021eaad8e5..9804e7f27b 100644
--- a/xbmc/music/windows/GUIWindowMusicNav.h
+++ b/xbmc/music/windows/GUIWindowMusicNav.h
@@ -33,6 +33,7 @@ protected:
   void PlayItem(int iItem) override;
   void OnWindowLoaded() override;
   void GetContextButtons(int itemNumber, CContextButtons &buttons) override;
+  bool OnPopupMenu(int iItem) override;
   bool OnContextButton(int itemNumber, CONTEXT_BUTTON button) override;
   bool OnClick(int iItem, const std::string &player = "") override;
   std::string GetStartFolder(const std::string &url) override;
diff --git a/xbmc/pvr/windows/GUIWindowPVRRecordings.cpp b/xbmc/pvr/windows/GUIWindowPVRRecordings.cpp
index 9edfb0f189..faae0580f8 100644
--- a/xbmc/pvr/windows/GUIWindowPVRRecordings.cpp
+++ b/xbmc/pvr/windows/GUIWindowPVRRecordings.cpp
@@ -121,6 +121,17 @@ bool CGUIWindowPVRRecordingsBase::OnAction(const CAction& action)
   return CGUIWindowPVRBase::OnAction(action);
 }
 
+bool CGUIWindowPVRRecordingsBase::OnPopupMenu(int iItem)
+{
+  if (iItem >= 0 && iItem < m_vecItems->Size())
+  {
+    const auto item = m_vecItems->Get(iItem);
+    item->SetProperty("CheckAutoPlayNextItem", {true});
+  }
+
+  return CGUIWindowPVRBase::OnPopupMenu(iItem);
+}
+
 bool CGUIWindowPVRRecordingsBase::OnContextButton(int itemNumber, CONTEXT_BUTTON button)
 {
   if (itemNumber < 0 || itemNumber >= m_vecItems->Size())
diff --git a/xbmc/pvr/windows/GUIWindowPVRRecordings.h b/xbmc/pvr/windows/GUIWindowPVRRecordings.h
index 5091fde2e7..21dc269772 100644
--- a/xbmc/pvr/windows/GUIWindowPVRRecordings.h
+++ b/xbmc/pvr/windows/GUIWindowPVRRecordings.h
@@ -30,6 +30,7 @@ public:
   bool OnMessage(CGUIMessage& message) override;
   bool OnAction(const CAction& action) override;
   void GetContextButtons(int itemNumber, CContextButtons& buttons) override;
+  bool OnPopupMenu(int iItem) override;
   bool OnContextButton(int itemNumber, CONTEXT_BUTTON button) override;
   bool Update(const std::string& strDirectory, bool updateFilterPath = true) override;
   void UpdateButtons() override;
diff --git a/xbmc/video/ContextMenus.cpp b/xbmc/video/ContextMenus.cpp
index 5f7ab86f60..ff988db801 100644
--- a/xbmc/video/ContextMenus.cpp
+++ b/xbmc/video/ContextMenus.cpp
@@ -185,15 +185,12 @@ void SetPathAndPlay(CFileItem& item)
   }
   else
   {
-    // auto play next item only works from inside certain media windows
-    ContentUtils::PlayMode mode = ContentUtils::PlayMode::PLAY_ONLY_THIS;
-    const int activeWindow = CServiceBroker::GetGUI()->GetWindowManager().GetActiveWindow();
-    if (activeWindow == WINDOW_VIDEO_NAV || activeWindow == WINDOW_TV_RECORDINGS ||
-        activeWindow == WINDOW_RADIO_RECORDINGS)
-      mode = ContentUtils::PlayMode::CHECK_AUTO_PLAY_NEXT_ITEM;
+    const ContentUtils::PlayMode mode = item.GetProperty("CheckAutoPlayNextItem").asBoolean()
+                                            ? ContentUtils::PlayMode::CHECK_AUTO_PLAY_NEXT_ITEM
+                                            : ContentUtils::PlayMode::PLAY_ONLY_THIS;
+    VIDEO_UTILS::PlayItem(std::make_shared<CFileItem>(item), mode);
 
     item.SetProperty("playlist_type_hint", PLAYLIST::TYPE_VIDEO);
-    VIDEO_UTILS::PlayItem(std::make_shared<CFileItem>(item), mode);
   }
 }
 } // unnamed namespace
diff --git a/xbmc/video/windows/GUIWindowVideoNav.cpp b/xbmc/video/windows/GUIWindowVideoNav.cpp
index 7e2c5cd5b6..8f798fda19 100644
--- a/xbmc/video/windows/GUIWindowVideoNav.cpp
+++ b/xbmc/video/windows/GUIWindowVideoNav.cpp
@@ -859,6 +859,17 @@ void CGUIWindowVideoNav::GetContextButtons(int itemNumber, CContextButtons &butt
   }
 }
 
+bool CGUIWindowVideoNav::OnPopupMenu(int iItem)
+{
+  if (iItem >= 0 && iItem < m_vecItems->Size())
+  {
+    const auto item = m_vecItems->Get(iItem);
+    item->SetProperty("CheckAutoPlayNextItem", {true});
+  }
+
+  return CGUIWindowVideoBase::OnPopupMenu(iItem);
+}
+
 bool CGUIWindowVideoNav::OnContextButton(int itemNumber, CONTEXT_BUTTON button)
 {
   CFileItemPtr item;
diff --git a/xbmc/video/windows/GUIWindowVideoNav.h b/xbmc/video/windows/GUIWindowVideoNav.h
index 61503ff726..fc025de77f 100644
--- a/xbmc/video/windows/GUIWindowVideoNav.h
+++ b/xbmc/video/windows/GUIWindowVideoNav.h
@@ -53,6 +53,7 @@ protected:
   virtual void PlayItem(int iItem);
   void OnDeleteItem(const CFileItemPtr& pItem) override;
   void GetContextButtons(int itemNumber, CContextButtons &buttons) override;
+  bool OnPopupMenu(int iItem) override;
   bool OnContextButton(int itemNumber, CONTEXT_BUTTON button) override;
   bool OnAddMediaSource() override;
   bool OnClick(int iItem, const std::string &player = "") override;
```

Reply:
Yeap, not pretty as the behaviour is "tied" to the popup menu action but at least removes the window check from the way. Later it can be improved or further worked :)
+1

Reply:
You'll probably get a different behaviour if started from json-rpc / remotes though :)

Reply:
Yeap also JSON depending on gui is wrong anyway